src/identity/configuration/IdentityProviderFactory.ts

```diff
@@ -153,6 +156,22 @@ export class IdentityProviderFactory implements ProviderFactory {
       id_token_signed_response_alg: this.jwtAlg,
     };
 
+    // This field will be used to store the associated WebID
+    // in "Clients" generated for client_credentials authentication.
+    config.extraClientMetadata = {
+      properties: [ CLIENT_CREDENTIALS_WEBID_KEY ],
+      validator(ctx, key, value, metadata): void {
+        // In practice this should never happen unless for some reason a client tries to add this during registration.
```

Comment:
I'm not sure what "this" is by reading the code (what is this callback for?)

```suggestion
        // In practice the client should never…
```

Reply:
This is going to be a long reply to a) explain what happens here, b) how this comment made me think that this PR introduces a giant security issue, c) how I discovered it might not be the case, and d) why an extra check should be added just to be sure. Big roller coaster of emotions.

Client metadata has a set of keys it is allowed to have. Clients being the things that do OIDC requests to our server. Extra keys can be added if we explicitly say what they are. This is what happens in the `properties` entry. We add a new metadata key so clients can store which WebID they are associated with. This only makes sense when using client_credentials as we are modifying the interpretation of Client a bit there. When a new credentials token gets created, we add such a "Client" to the database together with the key indicating which WebID it corresponds with so we can return the correct WebID in the access tokens that get returned when a user does a client_credentials request.

The `validator` function allows us to validate contents of these extra keys that we have defined. As mentioned above, this key only makes sense in the case of client_credentials requests, so we throw an error in case there is Client metadata that has a value for such a key, but does not have client_credentials as accepted grant. We add this grant when creating a token so that metadata is fine. Client metadata also gets created during dynamic registration. This is the "standard" OIDC flow that happens when using the authn client and the login window pops up. But in that case the Client metadata gets generated by the OIDC provider library and will not have the client_credentials grant. (Actually I'm not 100% sure if there is a way around this or not so this is another point for d) ).

The potential security risk I realized (besides additionally the one mentioned right above), is that Solid has another way of returning Client metadata: through the WebID of a client. The issue is that it is perfectly possible to add a client_secret and a CLIENT_CREDENTIALS_WEBID_KEY entry to your client WebID, and add client_credentials to your accepted grants. If you then do a client_credentials request with your secret to our server, we would accept the request (since the secret is correct) and return the WebID from the Client JSON, which is whatever WebID the owner of that Client JSON decided to put there! Meaning they could authenticate as any WebID of their choice on our server.

Now it turns out the above does not work because of the following line of code: https://github.com/CommunitySolidServer/CommunitySolidServer/blob/5e6c42e1f82f3474c64c8f43ef1827775ce8cfa7/src/identity/storage/WebIdAdapterFactory.ts#L79-L80
We explicitly say that no authentication is needed for such a client, so the OIDC provider will error because authorization was attempted. And it also doesn't work without the auth header.

All of these issues and risks can easily be fixed fortunately. What we have to do when we receive a client_credentials request is check if the account associated with the WebID we are going to return is associated with the client_id (being the token identifier). This is the same check we do when deleting the token. The one issue is that the `validator` function above has to be synchronous, but the function that actually adds the WebID to the access token can return a Promise so it can probably be added there.

Reply:
This has been resolved by changing how credential data is stored internally.